lxd/instance/instance_utils.go

```diff
@@ -111,6 +111,12 @@ func ValidConfig(sysOS *sys.OS, config map[string]string, expanded bool, instanc
 		return fmt.Errorf("nvidia.runtime is incompatible with privileged containers")
 	}
 
+	// Validate pinning strategy when limits.cpu specifies static pinning
```

Comment:
Thanks, this looks like the correct place to do composite config key validation.

However this seems to allow me to set the `limits.cpu` to include a comma and set `limits.cpu.pin_strategy` to `auto` in a profile (i.e when `instanceType == instancetype.Any`).

Is this correct?

So the scenarios we want to test are:

1. `limits.cpu.pin_strategy=auto` and `limits.cpu` set to static pinning in profile: Not allowed.
1. `limits.cpu.pin_strategy=auto` in profile and `limits.cpu` set to static pinning in instance config: Allowed (with `limits.cpu` taking precedence).
1. `limits.cpu` set to static pinning in profile and `limits.cpu.pin_strategy=auto` in instance config: Allowed (with the auto pin strategy taking precedence).
1. `limits.cpu.pin_strategy=auto` and `limits.cpu` set to static pinning in instance: Not allowed.
1. `limits.cpu.pin_strategy=auto` set in container config: Not allowed

Any others ive missed or got wrong in your view?

Reply:
> However this seems to allow me to set the `limits.cpu` to include a comma and set `limits.cpu.pin_strategy` to `auto` in a profile (i.e when `instanceType == instancetype.Any`).
> 
> Is this correct?

When `limits.cpu.pin_strategy` is set to `auto` in a profile, `limits.cpu` should take only take precedence in instance config - so no, this is not correct.

> 1. `limits.cpu.pin_strategy=auto` and `limits.cpu` set to static pinning in profile: Not allowed.
> 2. `limits.cpu.pin_strategy=auto` in profile and `limits.cpu` set to static pinning in instance config: Allowed (with `limits.cpu` taking precedence).
> 3. `limits.cpu` set to static pinning in profile and `limits.cpu.pin_strategy=auto` in instance config: Allowed (with the auto pin strategy taking precedence).
> 4. `limits.cpu.pin_strategy=auto` and `limits.cpu` set to static pinning in instance: Not allowed.
> 5. `limits.cpu.pin_strategy=auto` set in container config: Not allowed
> 
> Any others ive missed or got wrong in your view?

I think you've captured all cases above 👍.

Reply:
I've covered all of the cases except #3. If `limits.cpu` is set to static pinning, we actually don't want to skip over the segment that handles fixed CPU pinning/load balancing. There is already logic to prevent rebalancing for static pinning in `fillFixedInstances`. Therefore, I don't think `limits.cpu.pin_strategy=auto` at the local level should take precedence over `limits.cpu` at the profile level. Auto pinning without a cpu limit provides 1 vCPU by default, so if `limits.cpu.pin_strategy` is set to `auto` at the instance level, how would we handle allocating a fixed range of CPUs as per the profile config?

Reply:
> Therefore, I don't think `limits.cpu.pin_strategy=auto` at the local level should take precedence over `limits.cpu`

Yes that makes sense indeed, if there is a static pinning setting in the profile then that should take precedence over the local `limits.cpu.pin_strategy=auto`.